File: java/leetcode/array/beautiful_arrangement_ii.java
/**
 * 667. Beautiful Arrangement II
 * Given two integers n and k, you need to construct a list which contains n different positive integers ranging from 1 to n and obeys the following requirement: 
Suppose this list is [a1, a2, a3, ... , an], then the list [|a1 - a2|, |a2 - a3|, |a3 - a4|, ... , |an-1 - an|] has exactly k distinct integers.

If there are multiple answers, print any of them.

Example 1:

Input: n = 3, k = 1
Output: [1, 2, 3]
Explanation: The [1, 2, 3] has three different positive integers ranging from 1 to 3, and the [1, 1] has exactly 1 distinct integer: 1.
Example 2:

Input: n = 3, k = 2
Output: [1, 3, 2]
Explanation: The [1, 3, 2] has three different positive integers ranging from 1 to 3, and the [2, 1] has exactly 2 distinct integers: 1 and 2.
Note:

The n and k are in the range 1 <= k < n <= 104.
 */

public class beautiful_arrangement_ii {
    public int[] constructArray(int n, int k) {
        int[] ret = new int[n];
        ret[0] = 1;
        for (int i = 1, interval = k; i <= k; i++, interval--) {
            ret[i] = i % 2 == 1 ? ret[i-1] + interval : ret[i-1] - interval;
        }
        for (int i = k+1; i < n; i++) {
            ret[i] = i+1;
        }
        return ret;
    }

}

File: golang/leetcode/arrays/beautiful_arrangement_ii.go
package arrays

// 667. Beautiful Arrangement II

func constructArray(n int, k int) []int {
	ret := make([]int, n)
	ret[0] = 1
	for i, interval := 1, k; i <= k; i, interval = i+1, interval-1 {
		if i%2 == 1 {
			ret[i] = ret[i-1] + interval
		} else {
			ret[i] = ret[i-1] - interval
		}
	}
	for i := k + 1; i < n; i++ {
		ret[i] = i + 1
	}
	return ret
}
